main.cpp

```diff
@@ -184,6 +200,21 @@ struct DoubleType
 
     operator double() const { return *a; }
 
+    DoubleType& apply( std::function<DoubleType&( double& )> func )
+    {
+        if( func != nullptr )
+            func( *a );
+        return *this;
```

Comment:
`func()` returns `DoubleType&`, so you should return that object.   If you mentally step through the execution, you'll see that the returned object is `*this`.
- [ ] revise.